torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,812 @@
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
+                params.append(param)
+        self.params = params
+
+        # Shard module parameters in place
+        self._shard_parameters()
+
+        # Make sure all parameters are sharded.
+        for n, p in self.named_parameters():
+            if not hasattr(p, "_is_sharded"):
+                raise RuntimeError(f"found unsharded parameter: {n} ; {p.size()}")
+        self._reset_lazy_init()
+
+        # Enum to indicate if we're in the forward/backward pass, idle, etc.
+        self.training_state = TrainingState_.IDLE
+
+        # Flag to guard against preparing gradients multiple times per backward pass.
+        self._pre_backward_hook_has_run = False
+
+    # setting two factors 'self.gradient_predivide_factor'
+    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+    def _cast_buffers(
+        self, device: Optional[torch.device] = None, memo: Optional[Set] = None
+    ) -> None:
+        """Move all buffers to the given *device*.
+        If *device* are not given, then they will default to
+        ``self.compute_device``. In the
+        case of nested FSDP instances, we will respect the child instance's
+        ``compute_device`` configuration.
+        Args:
+            device (torch.device, Optional):
+                device to cast buffers to (defaults to compute_device)
+            memo (Set, Optional):
+                set of modules that have already been processed
+        """
+        if memo is None:
+            memo = set()
+        for module in self.modules():
+            if module is not self and isinstance(module, FullyShardedDataParallel):
+                # Allow any child FSDP instances to handle their own buffers.
+                module._cast_buffers(device=device, memo=memo)
+            elif module not in memo:
+                memo.add(module)
+                for name, buf in module.named_buffers(recurse=False):
+                    if buf is None:
+                        continue
+                    buf = buf.to(device=device or self.compute_device)
+                    setattr(module, name, buf)
+
+    @torch.no_grad()
+    def _shard_parameters(self) -> None:
+        """
+        At initialization we wrap a module with full parameters and shard the
+        parameters in-place. Sharding is implemented by viewing each parameter
+        as a 1D Tensor and retaining only a single slice, where the slice size
+        is determined by the number of data parallel workers.
+        After this initial sharding is complete, the user can initialize a
+        ``torch.optim.Optimizer`` in the usual way, i.e.::
```

Comment:
curious (no need to change, we can fix doc when this package becomes public), how does `i.e.::` render?